### ISSUE TITLE
Release Nuget Package

### ISSUE THREAD

**Issue by CoreyKaylor** [CoreyKaylor/Lightning.NET#5]
It would be preferable to consume this project through a nuget package. I can setup the automation for creating (manual for publish) the nuget package _if_ the rake build script is agreeable to the project.

**Comment by ilyalukyanov** [comment]
NuGet package would be very valuable contribution. Let's discuss the deatails of automation with rake in the corresponding issue #3

**Comment by CoreyKaylor** [comment]
I think we should be close to pull the trigger on this one. Maybe work out the file permissions issues on *nix machines and the other API reorganization going on and then I can help make this happen. Or at least explain how to do this step and let you do it yourself. It will pull from the artifacts on the build server to do the publish.

One other thing to note, nuget packages have a convention of being named the same as the dll. This is how I set it up, but wondered if the project name should change to avoid confusion? Your call, just calling it out.

**Comment by ilyalukyanov** [comment]
Corey, i agree. The branch with API reorganizations is now in pull request for review. This weekend i'll have some free time to spend and can look at this permissions issue and then merge the pull request.
I've never published a package in nuget so it's interesting for me to do it myself and get this experience.
Should we really follow the convention? As i can see not everyone does. One popular example is Json.NET. If we change name to LightningDB there could be a misunderstanding as it's native lmdb project name.

**Comment by CoreyKaylor** [comment]
I'm indifferent either way.

**Comment by CoreyKaylor** [comment]
Ripple used to have a working publish but the latest Nuget.Core broke that functionality. The best bet now is to download the produced artifact(s) from the build server and use Nuget.exe to do the publish. I wouldn't recommend going through the web UI.
After you have setup an API key for your account and set it up for Nuget.exe

Something like `NuGet Push YourDownloadedPackage.nupkg`

More help http://docs.nuget.org/docs/creating-packages/creating-and-publishing-a-package

**Comment by CoreyKaylor** [comment]
Also, if your choice is to bundle the lmdb.dll 64 bit version. I would recommend changing up the nuspec to add the reference element to prevent nuget from trying to reference the native dll.

**Comment by ilyalukyanov** [comment]
Thank your for the guide! I want to include a 32-bit version of lmdb.dll.
To do so i will add a line like `<file src="path\to\lmdb.dll" target="content" />` to the nuspec. Is this correct?
I don't want to deal with Mixed assemblies but still look on module initializer way. Both 32 and 64 bit versions could be embedded as a resources and then automatically extracted on assembly load according to the current platform. It's probably not a task for coming version release but for future one.

**Comment by Bobris** [comment]
32-bit version of lmdb.dll is nearly useless because it uses Memory mapped files so than size of database is limited by 1-2 GB and that's not much for database...

**Comment by ilyalukyanov** [comment]
I don't ignore 64-bit system. I'm just not sure that i can make it work well on 64-bit systems in a limited time i have)

**Comment by ilyalukyanov** [comment]
I've finally got it work on both platforms. Not tested very well yet, but unit tests are all green. I'll try to build a nuget package and fix permissions issue tomorrow. 
I don't like the way i've built 64bit native because of [this](https://github.com/ilyalukyanov/lightningdb-win/commit/ac3e0cd50f9af86cf66cd346f17769e4b54fd4ba). But it shouldn't break stability AFAIK,

**Comment by CoreyKaylor** [comment]
The rakefile already builds 32 bit and 64 bit dll's, just not using the VS solution.

**Comment by ilyalukyanov** [comment]
I should have started from rake except postponing it for the end) I'll play with it this week... Didn't have enough free time to finish it yesterday.

**Comment by ilyalukyanov** [comment]
Corey, i've played with rake. I see what you've told about. Under MinGW everything builds on both platforms out of the box. In parallel i still support VS builds which are different and don't compile without some changes. It's not related to building a nuget package so we leave it. 

I've finally fixed an issue with 'permission denied' on *nix and committed it to the bittness-support branch.
Now i want to enchance a rakefile so it could build native binaries for both 32 and 64 platforms (in a single target and if both are possible) and then include both into a nuget package. A wrapper now chooses at runtime which one to P/Invoke according to bitness of executing process.

**Comment by CoreyKaylor** [comment]
It is possible to build for both. Take a look at the Mac build to compare
the flag it uses.

**Comment by ilyalukyanov** [comment]
I've edited my comment but it's too late. I was imprecise, i meant, now we need to run rake twice for each platform. I want to add a task to build both in one time because both are required for a nuget package

**Comment by ilyalukyanov** [comment]
Now i'm using [MinGW-w64](http://mingw-w64.sourceforge.net/download.php) for cross compillation. To produce both 32 and 64 binaries (at least on windows) i had to install 2 mingw toolchains and to specify the exact compiler in the task (i686-w64-mingw32-gcc or x86_64-w64-mingw32-gcc). Looks ugly but this is what i've come to so far.

**Comment by ilyalukyanov** [comment]
Corey, please take a look on this commit. It produces correct nuget package with binaries for both platforms and build passess all the tests (on windows, not tested on other platforms). What do you think, is it ok for publishing?
I know, my rakefile looks bad. Maybe there is more elegant way to describe this build logic?

**Comment by CoreyKaylor** [comment]
I'm not a big fan of compiling a different name for the 64 vs. 32 bit. I understand the need to do both, but ideally we just generate to both the x86/Debug folder and the standard /Debug folder. Additionally, I am surprised that nuget doesn't already support x86 vs. x64. It looks like the current guidance is to release to separate nuget packages.

**Comment by ilyalukyanov** [comment]
I'm not a fan of it too but i'm for speed and portability. 
Having separate package for each platform is good practice but we lose portability. The developer should decide: bound an application with 32 bit arch or make it available only for 64 bit arch. 
I considered several methods to make library portable:
- Mixed mode assembly. Rejected it because of possible performance degradation and because such assemblies aren't well supported by Mono.
- Embedding both libraries as a resources and extracting appropriate one on module load according to the process bitness. Despite this method is cls compliant it's not supported by C# so IL injections are required which is unnecessary complexity. Also it would be hard to upgrade native code embedded into a library withoud recompilling whole solution.
- Keeping separate libraries (differently named as in our case). Ugly but simple and portable solution. Developer has enough control on dependencies.

I think i should add P/Invoke code against library named lmdb (in addition to lmdb32 and lmdb64 ones) to fallback to it if platform specific dll isn't present so developer could deploy to a single platform. Maybe i'll apply T4 templates for it to not repeat myself.

**Comment by CoreyKaylor** [comment]
I still think having multiple names adds too much complexity where it isn't needed. I think in the worst case scenario there could be a nuget powershell install and uninstall script to deal with the references accordingly and inside the package organized into /x86 and /x64 folders.

**Comment by Bobris** [comment]
Because this native dll is loaded on first method used. You can use polymorphism. Have both implementations each pointing to correctly named dll and instantiate only correct one depending on:

```
    public static bool IsX86Mode
    {
        get { return IntPtr.Size == 4; }
    }
```

**Comment by ilyalukyanov** [comment]
Yes, that was an idea [=)](https://github.com/ilyalukyanov/Lightning.NET/blob/bittness-support/src/LightningDB/Platform.cs)

**Comment by andy-uq** [comment]
I prefer `System.Environment.Is64BitProcess`

**Comment by ilyalukyanov** [comment]
After a long delay, now i'm PhD and finally got some free time to publish a package - https://www.nuget.org/packages/LightningDB/
